### ISSUE TITLE
Saga Data not keeping up

### ISSUE THREAD

**Issue by mkoziel2000** [rebus-org/Rebus#958]
I noticed in Rebus 6.5.5 that when I define a straight forward Saga with 4 messages fired in a chain, the messages can get processed before the SagaData has a chance to get updated and passed along.  This basically makes the SagaData as a holder of state between messages not very useful.  I don't remember seeing this problem in earlier versions of Rebus, although the transport layer we used in the earlier version was SQL Server.  I'm currently testing out the Kafka transport layer in the 6.5.5 version where I see the issue (the Saga storage is still in SQL Server).

Here is the basic Saga setup

- Publish Msg 1
- Consume Msg 1 (IsInitiatedBy)
    - Process Msg 1
    - Set flag in Saga Data that Msg 1 is processed
    - Publish Msg 2
- Consume Msg 2
    - Process Msg 2
    - Set flag in Saga Data that Msg 2 is processed
    - Publish Msg 3
- Consume Msg 3
    - Process Msg 3
    - Set flag in Saga Data that Msg 3 is processed
    - If all 3 messages where processed (looking at Saga Data), call the MarkAsComplete() method
 
What I should expect to see is that all the Sagas I pump into the system should make it to the MarkAsComplete() command.  Instead, its about 50% of the Sagas that get marked and 50% that never satisfy the IF condition when processing "Msg 3".  It seems to be timing related around how long it takes to consume and process the message relative to Rebus getting the SagaData internally updated after a previous message is done processing.  In the cases where I get a stale SagaData, there is a Conflict Resolution invoked by Rebus as the result of a Concurrency collision.  It seems the only way to overcome this problem is to add an artificial delay to the publishing of the next message in the Saga using deferLocal().

Its an arbitrary test, but I was basically verifying that Kafka as a transport (and subscriber) with SQL Server as the storage for Saga would work together and function properly when I ran into the odd behavior.  I don't remember seeing this issue when I used SQLServer for the transport (maybe because the mechanism is slow enough that the SagaData can get processed in time before the next message gets handled) - Or it might have still happened but it was very infrequent.

Is there a way to guarantee that the SagaData is not stale when doing chain-based message sequencing?  I suspect that a command&control pattern might be better suited to leveraging SagaData (A central handler that deals with the sequencing logic to determine which message to publish next based on the last message published)...but then, thinking about that pattern, it could also suffer from the same stale SagaData scenario since the worker handler has to publish a message to return control back to the controller handler before it exits its handler method.

Please advise.  I have to imagine people must be running into this issue when they are hooked up to fast transport systems like Kafka or Rabbit.

Here is an example Saga that I was using for the test:

    public class TestSaga : IdempotentSaga<TestSagaData>,
                            IAmInitiatedBy<KickoffSagaMessages>,
                            IHandleMessages<SagaMessageEarth>,
                            IHandleMessages<SagaMessageWind>,
                            IHandleMessages<SagaMessageFire>,
                            IHandleMessages<KickoffSagaMessages>,
                            IHandleMessages<IFailed<ISagaMessage>>
    {
        private IBus _bus;

        private static ILogger _logger = LogManager.GetCurrentClassLogger();

        public TestSaga(IBus bus)
        {
            _bus = bus ?? throw new ArgumentNullException();
        }

        protected override async Task ResolveConflict(TestSagaData otherSagaData)
        {
            this.Data.Task1Processed = this.Data.Task1Processed || otherSagaData.Task1Processed;
            this.Data.Task2Processed = this.Data.Task2Processed || otherSagaData.Task2Processed;
            this.Data.Task3Processed = this.Data.Task3Processed || otherSagaData.Task3Processed;
        }
        #pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
        public async Task Handle(SagaMessageEarth message)
        #pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
        {
            try
            {
                if (!this.Data.Task1Processed)
                {
                    _logger.Info("Processing Earth - {id}", this.Data.SagaInstanceId);
                    this.Data.stuffDone += "Earth;";
                    this.Data.Task1Processed = true;
                }
                await _bus.Publish(new SagaMessageWind()
                {
                    SagaInstanceId = message.SagaInstanceId
                }).ConfigureAwait(false);
                PossiblyDone();
                _logger.Info("Published Wind...Done Processing Earth - {id}", this.Data.SagaInstanceId);
            }
            catch (Exception e)
            {
                _logger.Error(e, "WHAT Earth? - {id}", this.Data.SagaInstanceId);
                throw;
            }
        }

        #pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
        public async Task Handle(SagaMessageWind message)
        #pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
        {
            try 
            {
                if (!this.Data.Task2Processed)
                {
                    _logger.Info("Processing Wind - {id}", this.Data.SagaInstanceId);
                    this.Data.stuffDone += "Wind;";
                    this.Data.Task2Processed = true;
                }
                await _bus.Publish(new SagaMessageFire()
                {
                    SagaInstanceId = message.SagaInstanceId
                }).ConfigureAwait(false);
                PossiblyDone();
                _logger.Info("Published Fire...Done Processing Wind - {id}", this.Data.SagaInstanceId);
            }
            catch (Exception e)
            {
                _logger.Error(e, "WHAT Wind? - {id}", this.Data.SagaInstanceId);
                throw;
            }
        }

        #pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
        public async Task Handle(SagaMessageFire message)
        #pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
        {
            try
            {
                if (!this.Data.Task3Processed)
                {
                    //throw new Exception("Not going to finish this");
                    _logger.Info("Processing Fire - {id}", this.Data.SagaInstanceId);
                    this.Data.stuffDone += "Fire;";
                    this.Data.Task3Processed = true;
                }
                PossiblyDone();
                _logger.Info("Done Processing Fire - {id}", this.Data.SagaInstanceId);
            }
            catch (Exception e)
            {
                _logger.Error(e, "WHAT Fire? - {id}", this.Data.SagaInstanceId);
                throw;
            }
        }

        private void PossiblyDone()
        {
            if (this.Data.Task1Processed && this.Data.Task2Processed && this.Data.Task3Processed)
            {
                _logger.Info("Completed everything for {id}: {msg}", this.Data.SagaInstanceId, this.Data.stuffDone);
                MarkAsComplete();
            }
            else
            {
                _logger.Info("NOT Completed everything for {id}: {task1},{task2},{task3}", this.Data.SagaInstanceId, this.Data.Task1Processed, this.Data.Task2Processed, this.Data.Task3Processed);
            }
        }

        public async Task Handle(KickoffSagaMessages message)
        {
            _logger.Info("Processing Kickoff - {id}", this.Data.SagaInstanceId);
            this.Data.SagaInstanceId = message.SagaInstanceId;
            this.Data.stuffDone += "Initiated;";
            _bus.Publish(new SagaMessageEarth()
            {
                SagaInstanceId = message.SagaInstanceId
            });
            _logger.Info("Published Earth....Done Processing Kickoff - {id}", this.Data.SagaInstanceId);
        }

        protected override void CorrelateMessages(ICorrelationConfig<TestSagaData> config)
        {
            //config.CorrelateHeader<ISagaMessage>("rbs2-corr-id", d => d.Id);
            config.Correlate<KickoffSagaMessages>(m => m.SagaInstanceId, d => d.SagaInstanceId);
            config.Correlate<SagaMessageFire>(m => m.SagaInstanceId, d => d.SagaInstanceId);
            config.Correlate<SagaMessageEarth>(m => m.SagaInstanceId, d => d.SagaInstanceId);
            config.Correlate<SagaMessageWind>(m => m.SagaInstanceId, d => d.SagaInstanceId);
            config.Correlate<IFailed<ISagaMessage>>(m => m.Message.SagaInstanceId, d => d.SagaInstanceId);
        }

        #pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
        public async Task Handle(IFailed<ISagaMessage> message)
        #pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
        {
            _logger.Error("Unable to handle the message of type {msgtype} with error message {errMsg}", message.Message.GetType().Name, message.ErrorDescription);
        }
    }

**Comment by mookid8000** [comment]
Hi, I've tried reproducing the issue [here](https://github.com/rebus-org/Rebus/blob/master/Rebus.Tests/Bugs/ReproduceFunnySagaIssue.cs).

I think your experience might have been caused by a missing `await` in this block of code
```csharp
    public async Task Handle(KickoffSagaMessages message)
    {
        _logger.Info("Processing Kickoff - {id}", this.Data.SagaInstanceId);
        this.Data.SagaInstanceId = message.SagaInstanceId;
        this.Data.stuffDone += "Initiated;";
        _bus.Publish(new SagaMessageEarth()
        {
            SagaInstanceId = message.SagaInstanceId
        });
        _logger.Info("Published Earth....Done Processing Kickoff - {id}", this.Data.SagaInstanceId);
    }
```
which should have read
```csharp
    public async Task Handle(KickoffSagaMessages message)
    {
        _logger.Info("Processing Kickoff - {id}", this.Data.SagaInstanceId);
        this.Data.SagaInstanceId = message.SagaInstanceId;
        this.Data.stuffDone += "Initiated;";
        // 👇😅
        await _bus.Publish(new SagaMessageEarth()
        {
            SagaInstanceId = message.SagaInstanceId
        });
        _logger.Info("Published Earth....Done Processing Kickoff - {id}", this.Data.SagaInstanceId);
    }
```

Could you try again with that added to your code?

(closing for now – feel free to pick it up again, if it still doesn't behave as expected)

**Comment by mkoziel2000** [comment]
I just checked the code.  The await was indeed missing in the code snippet that I put in this issue.  When I check my code, it was present.  I think I was just trying to figure out the behavior of the timing issue and you got the remnants of that.  I believe I isolated it down to the behavior of the Publish() method.  If you add a Thread.Sleep(100) right after the publish in one of the handlers, you will see the behavior unfold.  It seems that Rebus doesn't know that the next message pulled from the queue is the result of a publish() call that was made from within a handler that hasn't exited its method and saved its saga data yet.  Therefore, when the new message is associated with the same saga, the new saga data is not available so the old saga data is referenced.  This, in turn, results in the Conflict resolution event getting triggered.  Unfortunately, processing of that event may be too late as the message that is suppose to complete the saga may have already finished it processing by that point.  Unless the saga instance is available to the conflict resolution event such that we can mark it final, it seems we will end up with unfinished sagas.

I see how this can be the case given that everything is parallel and concurrent in nature.  The question is whether we can get rebus to persist the changed state of the saga data before the next message gets consumed by the next process/thread.  Seems like passing in the saga data as part of the publish() might be a good option to solve this issue as it gives Rebus the change to save off the change in saga data state before that message is pushed to the Q.  Another option could be that changes to Saga Data are only done through an accesssor, where the accessor will persist each change made...but that would probably break a lot of implementations out there.  Third option might be to create an Api that allows for registering publication calls with rebus to initiate as soon as the handler method is done...this would provide the hook necessary to get any change in Saga Data to get persisted first (this is close to the behavior of the DelayLocal() method...but that method is not predictable as to when it will fire relative to the current handler finishing).  Just some ideas.

**Comment by mookid8000** [comment]
Oh, sorry – could you remove all that `.ConfigureAwait(false)` too? I think it causes the bus operations to happen immediately and not be enlisted in the current handler transaction....

**Comment by mkoziel2000** [comment]
no change in behavior.  Curious why you might believe the ConfigureAwait(false) would affect this type of problem since all its doing is changing the SynchronizationContext that the rest of the code in the method runs under.  Could this be a problem at the transport connector level?  I guess I don't know how Saga handlers would infer synchronicity to make sure that a message (associated with the same saga) can get published by a handler and then wait until that handler exits before it places it onto the physical message queue.  I'm currently replacing Kafka with SQL Server to see if the problem "goes away".

**Comment by mookid8000** [comment]
>  guess I don't know how Saga handlers would infer synchronicity to make sure that a message (associated with the same saga) can get published by a handler and then wait until that handler exits before it places it onto the physical message queue.

Most of Rebus doesn't really know about sagas – it just happens to provide a little bit of niceness around loading/saving state, but the overall handling of messages ALWAYS goes like this:

1. Incoming message is "received" from queue (i.e. made unreceivable to other consumers)
2. User code is executed
3. Outgoing messages are sent
4. Incoming message is ACKed (i.e. actually removed from queue)

During step 2, "user code" (i.e. message handlers, in your case more specifically a saga handler) that sends/publishes messages get all send operations enlisted in Rebus' ambient transaction context, which is an `AsyncLocal<ITransactionContext>` that the bus automatically picks up if present.

If the bus does not find an ambient transaction context, it will surround every single send operation by its own little transaction context, meaning that it will happen immediately.

Honestly, I've never fully understood what `.ConfigureAwait(false)` ACTUALLY did, and what the purpose was. I just know that the use of it has been involved in several hard-to-track-down cases of users reporting funny Rebus behavior to me, and then their code was littered with this funny thing that they didn't understand why they used, except that "it's best practice to do it".

If removing it from your code and properly `await`ing your `Task`s didn't make a change, then maybe you could post some more information about your configuration? Specifically, I would be interested in seeing how you register the bus in your IoC container...

**Comment by mookid8000** [comment]
Ok. My suspicion is that this is due to a misbehaving Kafka transport, because it sounds like it sends messages right away. The correct behavior for any Rebus transport is to delay all send/publish operations until after the user code has been correctly executed (i.e. in your case: the saga data has been updated), and only THEN send/publish outgoing messages.

Please note that Rebus does not have an official Kafka transport, and there's good reasons why you would most likely NOT use a queue-oriented library like Rebus on top of a log-based broker.

**Comment by glazkovalex** [comment]
> My suspicion is that this is due to a misbehaving Kafka transport, because it sounds like it sends messages right away.

Hello! 
Sorry about the necropost, but no, it's not related to that. This could make a difference when throwing an exception when processing messages. But in this case there are no exceptions, all messages are processed correctly. By the way, starting from [version 3.0.1 of Rebus.Kafka](https://github.com/glazkovalex/Rebus.Kafka?tab=readme-ov-file#v-301-12012024), sending confirmation is postponed until message processing is completed, which provides support for the regular functionality of the Rebus bus: automatic retries and error handling. Which had to be worked out manually on earlier versions. 

@mkoziel2000 is absolutely right, it's about the delay in recording the saga in the database. The reply to the message arrives before the saga has been saved! That is, in this case, the problem is that saving saga data has nothing to do with sending messages! In a reliable and fault-tolerant implementation of the saga, the outbox pattern should be implemented! 
That is, Rebus must first save the current saga data and the messages that need to be sent at this step. And only then send messages. If successful, delete the data of the sent message from the outbox. And in case of failure to send messages, periodically try to send them when resuming communication with transport. Only in this case can the saga work reliable. Now, like @mkoziel2000, I am faced with the problem of using Rebus capabilities to implement processes (sagas) for reliable and fault-tolerant sagas, or switch to Zeebe, Temporary or DurableTask. I am sure that this typical Rebus application case with processes (sagas) was in demand by many.

@mookid8000, is there anywhere an example of using Rebus to implement reliable and fault-tolerant sagas with fast transport (RebbitMq or other fast transport) and saving state to a slow relational database (MSSQL, Postgresql or other relational database)?

I see it here [#819](https://github.com/rebus-org/Rebus/issues/819) something has been done to implement outbox. Also in the [Rebus.Outbox](https://github.com/rsivanov/Rebus.Outbox) package has also done something to implement outbox. @mookid8000, have you ever come across an example of implementing reliable and fault-tolerant sagins on Rebus along with outbox?

**Comment by mookid8000** [comment]
@glazkovalex You don't need an outbox with Rebus to make it buffer outgoing messages and thus execute stuff in the following order when processing messages in a saga:

1. Receive incoming message
2. Load saga data via correlation property
3. Execute saga handler - buffer outgoing messages
4. Update saga data
5. Send/publish outgoing messages
6. ACK incoming message

When executing actions in this order, it should ensure that e.g. requests sent from a saga do not result in suddenly receiving replies before even having saved the saga data.

If this is not what you're seeing with the Kafka transport, then I my suspicion is that it has not been correctly implemented, and I also don't believe it would pass the transport contract tests (which can be seen here, in case you're curious: https://github.com/rebus-org/Rebus/tree/master/Rebus.Tests.Contracts/Transports )

**Comment by glazkovalex** [comment]
Hello, @mookid8000! 
Thank you for the quick response and the recommendation of the transport tests! 
Yes, these tests do not pass due to the specifics of the transport implementation, but this does not mean that the Rebus.Kafka transport does not work in most Rebus bus modes. Perhaps someday I will develop the Rebus.Kafka transport to fully comply with these tests. But for now, I have plans to increase support for Kafka modes within the partially supported Rebus capabilities. Such as: support for Avro and Protobuf format; Support for schema registries; Support for key partitioning; The external setting of the initial offset for the queue and other important Apache Kafka functions. 
It's a pity that you didn't have an example of Rebus configuration at hand for a reliable, fault-tolerant Saga. 
Thanks for telling me that I don't need to add an "outbox". On older versions of the libraries, I reproduced the problem described by @mkoziel2000.  **I was wrong, the implementation of the Sagas in the Rebus is good, the sagas already have an implementation of the "outbox" pattern. Sagas did not work correctly on Rebus.Kafka transport because they rely on Rebus transcriptions, which were not previously supported in Rebus.Kafka.** 😕 When implementing the transport, I ignored all transactions, since Apache Kafka does not support any transactions. And it turned out that the "transactions" in Rebus are not transactions, but some kind of non-atomic delayed sending or not sending messages, which are actively used in Sagas.
By the way, Rebus.Kafka versions 3.0.1 and later supports these "transactions" and sending confirmations of receiving messages only after their successful processing.  
On the current version of Rebus 8.1 and Rebus.Kafka 3.2.1 transport, I selected the following set of parameters: 
```
builder.ConfigureServices((hostContext, services) =>
{
    var consoleLoggerFactory = new ConsoleLoggerFactory(true) { MinLevel = LogLevel.Debug };
    services.AddSingleton(consoleLoggerFactory);
    services.AutoRegisterHandlersFromAssemblyOf<TestSaga>();
    var consumerConfig = new ConsumerAndBehaviorConfig(kafkaEndpoint, "temp") { BehaviorConfig = new ConsumerBehaviorConfig { CommitPeriod = 1 } };
    services.AddRebus((configurer, serviceProvider) => configurer
        .Logging(l => l.Use(consoleLoggerFactory))
        .Transport(t => t.UseKafka(kafkaEndpoint, $"{nameof(IdempotentSaga)}.queue", new ProducerConfig(), consumerConfig))
        .Sagas(s => s.StoreInPostgres(connectionString, "SagasData", "SagasIndex", true, null, schemaName: "rebus"))
        .Timeouts(t => t.StoreInPostgres(new PostgresConnectionHelper(connectionString), "Timeouts", true, "rebus"))
        .Options(o =>
        {
            o.EnableIdempotentSagas();
            o.RetryStrategy($"{nameof(IdempotentSaga)}.queue.error", 3);
            o.UseAttributeOrTypeFullNameForTopicNames();
        })
    );
});
```
I hope it's the right one. 
Using [this example](https://github.com/glazkovalex/Rebus.Kafka/blob/master/Examples/IdempotentSaga/Program.cs), I reproduced @mkoziel2000's case. **I have checked the Sagas with various transport options many times - everything works correctly!** 
Having discovered your simplified version of the saga test without persistence, I added [it to my tests just in case](https://github.com/glazkovalex/Rebus.Kafka/blob/master/Rebus.Kafka.Tests/SagaTests.cs). Of course, I have it with Rebus.Kafka 3.2.1 transport also successfully passes. 
In terms of providing more complete support for the transport contract, I have a few small questions for you about the implementation of the contract. And also a couple of suggestions for improving the implementation of idempotent Sagas. I will ask you these questions in a separate issues, in the hope that you will help me with them a little.
